forc-pkg/src/source/ipfs.rs

```diff
@@ -0,0 +1,233 @@
+    }
+}
+
+impl Cid {
+    /// Using local node, fetches the content described by this cid.
+    async fn fetch_with_client(&self, ipfs_client: &IpfsClient, dst: &Path) -> Result<()> {
+        let cid_path = format!("/ipfs/{}", self.0);
+        // Since we are fetching packages as a fodler, they are returned as a tar archive.
+        let bytes = ipfs_client
+            .get(&cid_path)
+            .map_ok(|chunk| chunk.to_vec())
+            .try_concat()
+            .await?;
+        // After collecting bytes of the archive, we unpack it to the dst.
+        let mut archive = Archive::new(bytes.as_slice());
+        archive.unpack(dst)?;
+        Ok(())
+    }
+
+    /// Using a public gateway, fetches the content described by this cid.
+    async fn fetch_with_public_gateway(&self, dst: &Path) -> Result<()> {
+        let client = reqwest::Client::new();
+        // We request the content to be served to us in tar format by the public gateway.
+        let fetch_url = format!(
+            "{}/ipfs/{}?download=true&format=tar&filename={}.tar",
+            PUBLIC_GATEWAY, self.0, self.0
+        );
+        let req = client.get(fetch_url);
+        let res = req.send().await?;
+        let bytes: Vec<_> = res.text().await?.bytes().collect();
+
+        // After collecting bytes of the archive, we unpack it to the dst.
+        let mut archive = Archive::new(bytes.as_slice());
+        archive.unpack(dst)?;
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+pub enum PinnedParseError {
+    Prefix,
+    Cid(<cid::Cid as FromStr>::Err),
+}
+
+impl FromStr for Pinned {
+    type Err = PinnedParseError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        // ipfs+<cid>
+        let s = s.trim();
+        // Parse the prefix.
+        let prefix_plus = format!("{}+", Self::PREFIX);
+        if s.find(&prefix_plus) != Some(0) {
+            return Err(PinnedParseError::Prefix);
+        }
+        let s = &s[prefix_plus.len()..];
+        // Then the CID.
+        let cid: cid::Cid = s.parse().map_err(PinnedParseError::Cid)?;
+        Ok(Self(Cid(cid)))
+    }
+}
+
+impl Serialize for Cid {
+    fn serialize<S: Serializer>(&self, s: S) -> Result<S::Ok, S::Error> {
+        let cid_string: String = format!("{}", self.0);
+        cid_string.serialize(s)
+    }
+}
+
+impl<'de> Deserialize<'de> for Cid {
+    fn deserialize<D: Deserializer<'de>>(d: D) -> Result<Self, D::Error> {
+        use serde::de::Error;
+        let cid_string = String::deserialize(d)?;
+        let cid: cid::Cid = cid_string.parse().map_err(|e| {
+            let msg = format!("failed to parse CID from {cid_string:?}: {e}");
+            D::Error::custom(msg)
+        })?;
+        Ok(Self(cid))
+    }
+}
+
+fn ipfs_dir() -> PathBuf {
+    forc_util::user_forc_directory().join("ipfs")
+}
+
+fn cache_dir() -> PathBuf {
+    ipfs_dir().join("cache")
+}
+
+fn pkg_cache_dir(cid: &Cid) -> PathBuf {
+    cache_dir().join(format!("{}", cid.0))
+}
```

Comment:
These should maybe be placed in a separate file, or at least be at the top of this file so that it's easier to find